blockstore/splitstore/markset_badger.go

```diff
@@ -92,86 +68,123 @@ func (e *BadgerMarkSetEnv) Create(name string, sizeHint int64) (MarkSet, error)
 	return ms, nil
 }
 
+func (e *BadgerMarkSetEnv) Create(name string, sizeHint int64) (MarkSet, error) {
+	return e.create(name, sizeHint)
+}
+
+func (e *BadgerMarkSetEnv) CreateVisitor(name string, sizeHint int64) (MarkSetVisitor, error) {
+	return e.create(name, sizeHint)
+}
+
+func (e *BadgerMarkSetEnv) SupportsVisitor() bool { return true }
+
 func (e *BadgerMarkSetEnv) Close() error {
 	return os.RemoveAll(e.path)
 }
 
 func (s *BadgerMarkSet) Mark(c cid.Cid) error {
 	s.mx.Lock()
-
 	if s.pend == nil {
 		s.mx.Unlock()
 		return errMarkSetClosed
 	}
 
-	s.pend[string(c.Hash())] = struct{}{}
+	write, seqno := s.put(string(c.Hash()))
+	s.mx.Unlock()
 
-	if len(s.pend) < badgerMarkSetBatchSize {
-		s.mx.Unlock()
-		return nil
+	if write {
+		return s.write(seqno)
 	}
 
-	pend := s.pend
-	seqno := s.seqno
-	s.seqno++
-	s.writing[seqno] = pend
-	s.pend = make(map[string]struct{})
-	s.writers++
-	s.mx.Unlock()
+	return nil
+}
 
-	defer func() {
-		s.mx.Lock()
-		defer s.mx.Unlock()
+func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+	s.mx.RLock()
+	defer s.mx.RUnlock()
 
-		delete(s.writing, seqno)
-		s.writers--
-		if s.writers == 0 {
-			s.cond.Broadcast()
-		}
-	}()
+	key := c.Hash()
+	pendKey := string(key)
 
-	empty := []byte{} // not nil
+	has, err := s.tryPending(pendKey)
+	if has || err != nil {
+		return has, err
+	}
 
-	batch := s.db.NewWriteBatch()
-	defer batch.Cancel()
+	return s.tryDB(key)
+}
 
-	for k := range pend {
-		if err := batch.Set([]byte(k), empty); err != nil {
-			return err
+func (s *BadgerMarkSet) Visit(c cid.Cid) (bool, error) {
+	key := c.Hash()
+	pendKey := string(key)
+
+	s.mx.RLock()
+
+	has, err := s.tryPending(pendKey)
+	if has || err != nil {
+		s.mx.RUnlock()
+		return false, err
+	}
+
+	has, err = s.tryDB(key)
+	if has || err != nil {
+		s.mx.RUnlock()
+		return false, err
+	}
+
+	// we need to upgrade the lock to exclusive in order to write; take the version count to see
+	// if there was another write while we were upgrading
+	version := s.version
+	s.mx.RUnlock()
+
+	s.mx.Lock()
+	// we have to do the check dance again
+	has, err = s.tryPending(pendKey)
+	if has || err != nil {
+		s.mx.Unlock()
+		return false, err
+	}
+
+	if version != s.version {
+		// something was written to the db, we need to check it
+		has, err = s.tryDB(key)
+		if has || err != nil {
+			s.mx.Unlock()
+			return false, err
```

Comment:
I'd still consider that code I posted as it should simplify this function a bit (even if you can't defer). But that's not critical.

Reply:
yeah, very small delta; I'll leave it as is.